### PR TITLE
Muon

### PR DESCRIPTION
**Pull request by arthurcolle**

### CODE REVIEW

**Comment by N8python** [comment]
This is a MASSIVE PR. Could you explain why you want to change so much + what each part does?

**Comment by N8python** [comment]
Because the amount of work you did is HUGE - extreme kudos for working out an entire MLX-CUDA parallel training pipeline.

**Comment by N8python** [comment]
This is an extremely amazing PR. Kudos for working out all of this distributed training infra - how did you do it so fast? I'm not sure whether it should be merged into the main repo, however, as this is a local, mlx, pretraining library and not an mlx + cuda library. If you could maybe organize the cuda stuff (and all the shell descripts) into their own folder away from the main MLX stuff as a clear add-on, merging could make sense. Either way, I'll link to your repo in the README.

**Comment by N8python** [comment]
I try ' ./run_256m_distributed.sh' but the follow error is encountered:

```
--
=> Step 2: RUN DEBIAN_FRONTEND=noninteractive apt-get install -y libnccl2 libnccl-dev
Apr 26  11:10:26.345 | Reading package lists...
Apr 26  11:10:26.457 | Building dependency tree... Reading state information...E
Apr 26  11:10:26.462 | : Unable to locate package libnccl2 E: Unable to locate package libnccl-dev
Apr 26  11:10:26.748 | Terminating task due to error: failed to run builder command "DEBIAN_FRONTEND=noninteractive apt-get install -y libnccl2 libnccl-dev": container exit status: 100
Apr 26  11:10:26.834 | Runner failed with exception: task exited with failure, status = exit status: 1
Apr 26  11:10:26.867 | Stopping app - uncaught exception raised locally: RemoteError('Image build for im-YkC8vhXtEHMlgsj1fa6aD0 failed with the exception:\ntask exited with failure, status = exit status: 1').
 

<br class="Apple-interchange-newline">
```

**Comment by arthurcolle** [comment]
@N8python I really appreciate the kind feedback. I love mlx, and I up until recently had two M2 Max 64GB nodes, and was able to run inference using ExoLabs' Exo distributed inference repo. I no longer have two M2 Max nodes, but I used Modal.com extensively last year, so I wanted to see what it would take to actually distribute the workload across some remote worker nodes on Modal.com, while doing the gradient accumulation piece on my local machine - mostly just for learning, and to be able to wrap parts of the training code with visualization or other aggregate metadata collection. Really interested in multi-modal pretraining and adding in labelled image datasets was another thing I looked at yesterday.

I kind of closed the first one because seemed excessively crazy, but I thought the Muon implementation was really nice and it runs extremely smoothly for me. I haven't actually been able to get the distributed piece fully working, but I just wanted to share some of this. Thanks again for the feedback, let me take a look at the bug you shared + clean up the mess.

**Comment by N8python** [comment]
Got it - once you get the distributed stuff fully working, I'll take a look and consider how we could proceed with merging or integrating this.

**Comment by N8python** [comment]
And how does your muon differ from what we already have in mlx-optimizers?
